<a id="setup"></a>
# <p style="background-color: #ff6200; font-family:calibri; color:white; font-size:120%; font-family:Verdana; text-align:center; border-radius:15px 50px;">Step 1 | Setup and Initialization</p>

<div style="border-radius:10px; padding: 15px; font-size:100%; text-align:left">

<h2 align="left"><font color=#ff6200>Libraries</font></h2>

<div style="border-radius:10px; padding: 15px;font-size:100%; text-align:left">

<h2 align="left"><font color=#ff6200>Dataset Description</font></h2>

| __Variable__   | __Description__ |
|     :---       |       :---      |      
| __ID__  | The identifier. |
| __brand__  | The brand of the car. |
| __model__| The model of the car. |
| __model_year__   | The year of the model. |
| __milage__| How many miles the car have. |
| __fuel_type__  | The type of fuel. |
| __engine__ | Engine details. |
| __transmission__    |Transmission details. |
| __ext_col__    | Exterior color. |
| __int_col__    | Interior color. |
| __accident__    | The car has an accident. (1 if yes) |
| __clean_title__    | Clean Tittle. |
| __price__    | The price of car. (Target Variable) |


In [1]:
import polars as pl
import polars.selectors as cs

In [2]:
df = (
    pl.read_csv("train.csv")
)

<a id="setup"></a>
# <p style="background-color: #ff6200; font-family:calibri; color:white; font-size:120%; font-family:Verdana; text-align:center; border-radius:15px 50px;">Step 2 | EDA</p>


<div style="border-radius:10px; padding: 15px; background-color: #ffeacc; font-size:100%; text-align:left">
Lets do an Exploratory Data Analysis (EDA) over the dataset to get a better understanding.    
</div>


In [3]:
df.head()

shape: (5, 13)
┌─────┬──────────────┬──────────────┬────────────┬───┬─────────┬─────────────┬─────────────┬───────┐
│ id  ┆ brand        ┆ model        ┆ model_year ┆ … ┆ int_col ┆ accident    ┆ clean_title ┆ price │
│ --- ┆ ---          ┆ ---          ┆ ---        ┆   ┆ ---     ┆ ---         ┆ ---         ┆ ---   │
│ i64 ┆ str          ┆ str          ┆ i64        ┆   ┆ str     ┆ str         ┆ str         ┆ i64   │
╞═════╪══════════════╪══════════════╪════════════╪═══╪═════════╪═════════════╪═════════════╪═══════╡
│ 0   ┆ MINI         ┆ Cooper S     ┆ 2007       ┆ … ┆ Gray    ┆ None        ┆ Yes         ┆ 4200  │
│     ┆              ┆ Base         ┆            ┆   ┆         ┆ reported    ┆             ┆       │
│ 1   ┆ Lincoln      ┆ LS V8        ┆ 2002       ┆ … ┆ Beige   ┆ At least 1  ┆ Yes         ┆ 4999  │
│     ┆              ┆              ┆            ┆   ┆         ┆ accident or ┆             ┆       │
│     ┆              ┆              ┆            ┆   ┆         ┆ damage …    ┆             ┆       │
│ 2   ┆ Chevrolet    ┆ Silverado    ┆ 2002       ┆ … ┆ Gray    ┆ None        ┆ Yes         ┆ 13900 │
│     ┆              ┆ 2500 LT      ┆            ┆   ┆         ┆ reported    ┆             ┆       │
│ 3   ┆ Genesis      ┆ G90 5.0      ┆ 2017       ┆ … ┆ Black   ┆ None        ┆ Yes         ┆ 45000 │
│     ┆              ┆ Ultimate     ┆            ┆   ┆         ┆ reported    ┆             ┆       │
│ 4   ┆ Mercedes-Ben ┆ Metris Base  ┆ 2021       ┆ … ┆ Beige   ┆ None        ┆ Yes         ┆ 97500 │
│     ┆ z            ┆              ┆            ┆   ┆         ┆ reported    ┆             ┆       │
└─────┴──────────────┴──────────────┴────────────┴───┴─────────┴─────────────┴─────────────┴───────┘

<div style="border-radius:10px; padding: 15px; font-size:100%; text-align:left">

<h2 align="left"><font color=#ff6200>String Variables</font></h2>

In [4]:
(
    df.select(cs.by_dtype(pl.String))
    .head(5)
)

shape: (5, 9)
┌────────────┬────────────┬────────────┬───────────┬───┬─────────┬─────────┬───────────┬───────────┐
│ brand      ┆ model      ┆ fuel_type  ┆ engine    ┆ … ┆ ext_col ┆ int_col ┆ accident  ┆ clean_tit │
│ ---        ┆ ---        ┆ ---        ┆ ---       ┆   ┆ ---     ┆ ---     ┆ ---       ┆ le        │
│ str        ┆ str        ┆ str        ┆ str       ┆   ┆ str     ┆ str     ┆ str       ┆ ---       │
│            ┆            ┆            ┆           ┆   ┆         ┆         ┆           ┆ str       │
╞════════════╪════════════╪════════════╪═══════════╪═══╪═════════╪═════════╪═══════════╪═══════════╡
│ MINI       ┆ Cooper S   ┆ Gasoline   ┆ 172.0HP   ┆ … ┆ Yellow  ┆ Gray    ┆ None      ┆ Yes       │
│            ┆ Base       ┆            ┆ 1.6L 4    ┆   ┆         ┆         ┆ reported  ┆           │
│            ┆            ┆            ┆ Cylinder  ┆   ┆         ┆         ┆           ┆           │
│            ┆            ┆            ┆ Engine…   ┆   ┆         ┆         ┆           ┆           │
│ Lincoln    ┆ LS V8      ┆ Gasoline   ┆ 252.0HP   ┆ … ┆ Silver  ┆ Beige   ┆ At least  ┆ Yes       │
│            ┆            ┆            ┆ 3.9L 8    ┆   ┆         ┆         ┆ 1         ┆           │
│            ┆            ┆            ┆ Cylinder  ┆   ┆         ┆         ┆ accident  ┆           │
│            ┆            ┆            ┆ Engine…   ┆   ┆         ┆         ┆ or damage ┆           │
│            ┆            ┆            ┆           ┆   ┆         ┆         ┆ …         ┆           │
│ Chevrolet  ┆ Silverado  ┆ E85 Flex   ┆ 320.0HP   ┆ … ┆ Blue    ┆ Gray    ┆ None      ┆ Yes       │
│            ┆ 2500 LT    ┆ Fuel       ┆ 5.3L 8    ┆   ┆         ┆         ┆ reported  ┆           │
│            ┆            ┆            ┆ Cylinder  ┆   ┆         ┆         ┆           ┆           │
│            ┆            ┆            ┆ Engine…   ┆   ┆         ┆         ┆           ┆           │
│ Genesis    ┆ G90 5.0    ┆ Gasoline   ┆ 420.0HP   ┆ … ┆ Black   ┆ Black   ┆ None      ┆ Yes       │
│            ┆ Ultimate   ┆            ┆ 5.0L 8    ┆   ┆         ┆         ┆ reported  ┆           │
│            ┆            ┆            ┆ Cylinder  ┆   ┆         ┆         ┆           ┆           │
│            ┆            ┆            ┆ Engine…   ┆   ┆         ┆         ┆           ┆           │
│ Mercedes-B ┆ Metris     ┆ Gasoline   ┆ 208.0HP   ┆ … ┆ Black   ┆ Beige   ┆ None      ┆ Yes       │
│ enz        ┆ Base       ┆            ┆ 2.0L 4    ┆   ┆         ┆         ┆ reported  ┆           │
│            ┆            ┆            ┆ Cylinder  ┆   ┆         ┆         ┆           ┆           │
│            ┆            ┆            ┆ Engine…   ┆   ┆         ┆         ┆           ┆           │
└────────────┴────────────┴────────────┴───────────┴───┴─────────┴─────────┴───────────┴───────────┘

In [5]:
(
    df
    .select(cs.by_dtype(pl.String))
    .unpivot(variable_name="column")
    .group_by("column")
    .agg(
        Null_Values = (pl.col.value.is_null().sum() / pl.col.value.len()) * 100,
        Unique_Values = (pl.col.value.n_unique())
    )
)

shape: (9, 3)
┌──────────────┬─────────────┬───────────────┐
│ column       ┆ Null_Values ┆ Unique_Values │
│ ---          ┆ ---         ┆ ---           │
│ str          ┆ f64         ┆ u32           │
╞══════════════╪═════════════╪═══════════════╡
│ engine       ┆ 0.0         ┆ 1117          │
│ ext_col      ┆ 0.0         ┆ 319           │
│ transmission ┆ 0.0         ┆ 52            │
│ fuel_type    ┆ 2.69608     ┆ 8             │
│ accident     ┆ 1.300568    ┆ 3             │
│ int_col      ┆ 0.0         ┆ 156           │
│ brand        ┆ 0.0         ┆ 57            │
│ model        ┆ 0.0         ┆ 1897          │
│ clean_title  ┆ 11.360876   ┆ 2             │
└──────────────┴─────────────┴───────────────┘

<div style="border-radius: 15px; padding: 20px; background-color: #ffeacc; font-size: 100%; text-align: left; border: 2px solid #ff6200; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);">
    <h2 style="font-family: Arial, sans-serif; color: #ff6200; margin-bottom: 10px;">String Variables</h2>
    <ul style="font-family: Arial, sans-serif; line-height: 1.8; color: #333; padding-left: 20px;">
        <li><strong>Missing Values:</strong> Missing data is found in 3 variables: <em>"fuel_type"</em> (2.69%), <em>"clean_title"</em> (11.3%), and <em>"accident"</em> (1.3%). The <em>"clean_title"</em> variable has a significant portion of missing values, which I will address later in the process.
        </li>
        <li><strong>Unique Values:</strong> The dataset contains a high number of unique values, particularly in the <em>"model"</em> column, with 1,897 distinct entries.</li>        
        <li><strong>Brand:</strong> There are 57 different car brands in the dataset. I will categorize these into broader groups such as luxury cars, premium cars, mainstream brands, and others.</li>
        <li><strong>Fuel Type:</strong> The <em>"fuel_type"</em> column has 8 different values. To simplify the analysis, I will condense these into 5 categories: 1) Diesel, 2) Hybrid, 3) Electric, 4) Gasoline, and 5) Other.</li>
        <li><strong>Interior and Exterior Color:</strong> Although there are many color options, I will focus on the base colors and create a new column labeled "Special Color" to flag any unique color variations.</li>
        <li><strong>Engine:</strong> The <em>"engine"</em> column contains valuable information such as horsepower (HP), engine displacement (liters), and cylinder count.</li>
        <li><strong>Transmission:</strong> The <em>"transmission"</em> column provides interesting details. I will perform additional data wrangling to extract and organize the key information from this column.</li>
        <li><strong>Model:</strong> The <em>"Model"</em> column have multiple values, so on this first go, i'm going to ignore it.</li>
    </ul>
</div>


<div style="border-radius:10px; padding: 15px; font-size:100%; text-align:left">
<h3 align="left"><font color=#ff6200>Brand</font></h3>
</div>

In [6]:
# Brand
Luxury_list = ["Lambo", "Ferrari", "Bugatti", "Rolls-Royce", "Maserati", "McLaren", "Bentley", "Aston", "Lamborghini", "Maybach"]
Premium_list = ["Mercedes-Benz", "Audi", "BMW", "Porsche", "MINI", "Cadillac","Lexus", "Tesla", "INFINITI",
                "Land", "Lincoln", "Volvo", "Acura", "Jeep", "Honda", "Chrysler", "Genesis", "Pontiac", "Jaguar",
                "Hummer", "Alfa", "Dodge", "Subaru", "FIAT"]
Normal_list = ["Toyota", "Nissan", "Volkswagen", "Hyundai", "Kia", "Chevrolet", "Ford", "GMC", "Mazda", "RAM", "Mitsubishi", "Suzuki"]


def set_brand(brand_col: pl.Expr) -> pl.Expr:
    return (
        pl.when(brand_col.is_in(Luxury_list)).then(3)
        .when(brand_col.is_in(Premium_list)).then(2)
        .when(brand_col.is_in(Normal_list)).then(1)
        .otherwise(0)
        .alias("Brand Type")
    )


<div style="border-radius:10px; padding: 15px; font-size:100%; text-align:left">
<h3 align="left"><font color=#ff6200>Fuel</font></h3>
</div>

In [7]:
def set_fuel_type(fuel_type_col: pl.Expr) -> pl.Expr:
    return (
        pl.when(fuel_type_col.is_in(["E85 Flex Fuel", "Gasoline"])).then(1)
        .when(fuel_type_col == "Diesel").then(2)
        .when(fuel_type_col.is_in([ "Hybrid","Plug-In Hybrid" ])).then(3)
        .when(fuel_type_col.is_in([ "not supported"])).then(4)
        .when(fuel_type_col.is_in(["–"])).then(5)
        .otherwise(6)
        .alias("Fuel Type")
    )

<div style="border-radius:10px; padding: 15px; font-size:100%; text-align:left">
<h3 align="left"><font color=#ff6200>Engine</font></h3>
</div>

In [20]:
def get_engine(engine_col: pl.Expr) -> pl.Expr:
    return (
        pl.col("engine").str.extract(r'(\d+)\.\d+HP',1).alias("HP"),
        pl.col("engine").str.extract(r'(\d+\.\d+)L',1).alias("Displacement"),
        pl.col("engine").str.extract(r'(\d+)\sCylinder',1).alias("Cylinder"),
        pl.col("engine").str.contains("Electric").alias("Electric")
    )

In [21]:
(
    df
    .with_columns(
        get_engine(engine_col=pl.col("engine"))
    )
)

shape: (188_533, 17)
┌────────┬──────────────┬──────────────┬────────────┬───┬──────┬─────────────┬──────────┬──────────┐
│ id     ┆ brand        ┆ model        ┆ model_year ┆ … ┆ HP   ┆ Displacemen ┆ Cylinder ┆ Electric │
│ ---    ┆ ---          ┆ ---          ┆ ---        ┆   ┆ ---  ┆ t           ┆ ---      ┆ ---      │
│ i64    ┆ str          ┆ str          ┆ i64        ┆   ┆ str  ┆ ---         ┆ str      ┆ bool     │
│        ┆              ┆              ┆            ┆   ┆      ┆ str         ┆          ┆          │
╞════════╪══════════════╪══════════════╪════════════╪═══╪══════╪═════════════╪══════════╪══════════╡
│ 0      ┆ MINI         ┆ Cooper S     ┆ 2007       ┆ … ┆ 172  ┆ 1.6         ┆ 4        ┆ false    │
│        ┆              ┆ Base         ┆            ┆   ┆      ┆             ┆          ┆          │
│ 1      ┆ Lincoln      ┆ LS V8        ┆ 2002       ┆ … ┆ 252  ┆ 3.9         ┆ 8        ┆ false    │
│ 2      ┆ Chevrolet    ┆ Silverado    ┆ 2002       ┆ … ┆ 320  ┆ 5.3         ┆ 8        ┆ false    │
│        ┆              ┆ 2500 LT      ┆            ┆   ┆      ┆             ┆          ┆          │
│ 3      ┆ Genesis      ┆ G90 5.0      ┆ 2017       ┆ … ┆ 420  ┆ 5.0         ┆ 8        ┆ false    │
│        ┆              ┆ Ultimate     ┆            ┆   ┆      ┆             ┆          ┆          │
│ 4      ┆ Mercedes-Ben ┆ Metris Base  ┆ 2021       ┆ … ┆ 208  ┆ 2.0         ┆ 4        ┆ false    │
│        ┆ z            ┆              ┆            ┆   ┆      ┆             ┆          ┆          │
│ …      ┆ …            ┆ …            ┆ …          ┆ … ┆ …    ┆ …           ┆ …        ┆ …        │
│ 188528 ┆ Cadillac     ┆ Escalade ESV ┆ 2017       ┆ … ┆ 420  ┆ 6.2         ┆ 8        ┆ false    │
│        ┆              ┆ Platinum     ┆            ┆   ┆      ┆             ┆          ┆          │
│ 188529 ┆ Mercedes-Ben ┆ AMG C 43 AMG ┆ 2018       ┆ … ┆ 385  ┆ 3.0         ┆ 6        ┆ false    │
│        ┆ z            ┆ C 43 4MATIC  ┆            ┆   ┆      ┆             ┆          ┆          │
│ 188530 ┆ Mercedes-Ben ┆ AMG GLC 63   ┆ 2021       ┆ … ┆ 469  ┆ 4.0         ┆ 8        ┆ false    │
│        ┆ z            ┆ Base 4MATIC  ┆            ┆   ┆      ┆             ┆          ┆          │
│ 188531 ┆ Audi         ┆ S5 3.0T      ┆ 2022       ┆ … ┆ null ┆ 3.0         ┆ null     ┆ false    │
│        ┆              ┆ Prestige     ┆            ┆   ┆      ┆             ┆          ┆          │
│ 188532 ┆ Porsche      ┆ Macan Base   ┆ 2016       ┆ … ┆ 252  ┆ 2.0         ┆ 4        ┆ false    │
└────────┴──────────────┴──────────────┴────────────┴───┴──────┴─────────────┴──────────┴──────────┘

<div style="border-radius:10px; padding: 15px; font-size:100%; text-align:left">
<h3 align="left"><font color=#ff6200>Color</font></h3>
</div>

In [39]:
color_list = ["yellow","white", "silver", "black", "blue", "gray", "red", "green", "beige"]
pattern = "|".join(color_list)

(
    df
    .with_columns(
        pl.col("ext_col").str.strip_chars().str.to_lowercase(),
        pl.col("int_col").str.strip_chars().str.to_lowercase()
    )
    .with_columns(
        pl.col("ext_col").str.extract(f"({pattern})").alias("ext_color_f"),
        pl.col("int_col").str.extract(f"({pattern})").alias("int_col_f")
    )
    .filter(
        ~pl.col("ext_col").is_in(color_list)
    )

)


shape: (24_521, 15)
┌────────┬────────────┬────────────┬────────────┬───┬────────────┬────────┬────────────┬───────────┐
│ id     ┆ brand      ┆ model      ┆ model_year ┆ … ┆ clean_titl ┆ price  ┆ ext_color_ ┆ int_col_f │
│ ---    ┆ ---        ┆ ---        ┆ ---        ┆   ┆ e          ┆ ---    ┆ f          ┆ ---       │
│ i64    ┆ str        ┆ str        ┆ i64        ┆   ┆ ---        ┆ i64    ┆ ---        ┆ str       │
│        ┆            ┆            ┆            ┆   ┆ str        ┆        ┆ str        ┆           │
╞════════╪════════════╪════════════╪════════════╪═══╪════════════╪════════╪════════════╪═══════════╡
│ 8      ┆ Ford       ┆ F-150 XLT  ┆ 2020       ┆ … ┆ Yes        ┆ 62890  ┆ white      ┆ black     │
│ 20     ┆ Land       ┆ Rover      ┆ 2023       ┆ … ┆ null       ┆ 41998  ┆ black      ┆ black     │
│        ┆            ┆ Defender S ┆            ┆   ┆            ┆        ┆            ┆           │
│ 27     ┆ Genesis    ┆ G90 5.0    ┆ 2017       ┆ … ┆ Yes        ┆ 38000  ┆ null       ┆ black     │
│        ┆            ┆ Ultimate   ┆            ┆   ┆            ┆        ┆            ┆           │
│ 39     ┆ Buick      ┆ Enclave    ┆ 2022       ┆ … ┆ Yes        ┆ 63900  ┆ null       ┆ null      │
│        ┆            ┆ Premium    ┆            ┆   ┆            ┆        ┆            ┆           │
│ 43     ┆ Mercedes-B ┆ AMG GT 53  ┆ 2022       ┆ … ┆ Yes        ┆ 244896 ┆ black      ┆ black     │
│        ┆ enz        ┆ Base       ┆            ┆   ┆            ┆        ┆            ┆           │
│ …      ┆ …          ┆ …          ┆ …          ┆ … ┆ …          ┆ …      ┆ …          ┆ …         │
│ 188509 ┆ Ford       ┆ Explorer   ┆ 2017       ┆ … ┆ Yes        ┆ 26950  ┆ silver     ┆ null      │
│        ┆            ┆ Platinum   ┆            ┆   ┆            ┆        ┆            ┆           │
│ 188511 ┆ Land       ┆ Rover      ┆ 2022       ┆ … ┆ null       ┆ 23739  ┆ red        ┆ null      │
│        ┆            ┆ Range      ┆            ┆   ┆            ┆        ┆            ┆           │
│        ┆            ┆ Rover      ┆            ┆   ┆            ┆        ┆            ┆           │
│        ┆            ┆ Velar P380 ┆            ┆   ┆            ┆        ┆            ┆           │
│        ┆            ┆ S…         ┆            ┆   ┆            ┆        ┆            ┆           │
│ 188513 ┆ Lexus      ┆ GS 350 F   ┆ 2022       ┆ … ┆ null       ┆ 48990  ┆ null       ┆ black     │
│        ┆            ┆ Sport      ┆            ┆   ┆            ┆        ┆            ┆           │
│ 188515 ┆ GMC        ┆ Sierra     ┆ 2023       ┆ … ┆ Yes        ┆ 48999  ┆ white      ┆ null      │
│        ┆            ┆ 1500       ┆            ┆   ┆            ┆        ┆            ┆           │
│        ┆            ┆ Denali     ┆            ┆   ┆            ┆        ┆            ┆           │
│        ┆            ┆ Ultimate   ┆            ┆   ┆            ┆        ┆            ┆           │
│ 188531 ┆ Audi       ┆ S5 3.0T    ┆ 2022       ┆ … ┆ null       ┆ 84900  ┆ gray       ┆ black     │
│        ┆            ┆ Prestige   ┆            ┆   ┆            ┆        ┆            ┆           │
└────────┴────────────┴────────────┴────────────┴───┴────────────┴────────┴────────────┴───────────┘

In [52]:
df = pl.DataFrame(
    {
        "engine": ["172.0HP 1.6L 4 Cylinder Engine Gasoline Fuel",
                    "3.0 Liter Twin Turbo",
                    "429.0HP 5.0L 8 Cylinder Engine Gasoline Fuel"],
    }
)

(
    df
    .with_columns(
        pl.col("engine").str.replace(r'\sLiter', "L")
    )
    .with_columns(
        pl.col("engine").str.extract(r'(\d+)\.\d+HP',1).alias("HP"),
        pl.col("engine").str.extract(r'(\d+\.\d+)L',1).alias("Displacement"),
        pl.col("engine").str.extract(r'(\d+)\sCylinder',1).alias("Cylinder"),
    )
)

def get_engine(engine_col: pl.Expr) -> pl.Expr:
    return (
        pl.col("engine").str.extract(r'(\d+)\.\d+HP',1).alias("HP"),
        pl.col("engine").str.extract(r'(\d+\.\d+)L',1).alias("Displacement"),
        pl.col("engine").str.extract(r'(\d+)\sCylinder',1).alias("Cylinder"),
        pl.col("engine").str.contains("Electric").alias("Electric")
    )


(
    df
    .with_columns(
        pl.col("engine").str.replace(r'\sLiter', "L")
    )
    .with_columns(
        get_engine(pl.col("engine"))
    )
)

shape: (3, 5)
┌─────────────────────────────────┬──────┬──────────────┬──────────┬──────────┐
│ engine                          ┆ HP   ┆ Displacement ┆ Cylinder ┆ Electric │
│ ---                             ┆ ---  ┆ ---          ┆ ---      ┆ ---      │
│ str                             ┆ str  ┆ str          ┆ str      ┆ bool     │
╞═════════════════════════════════╪══════╪══════════════╪══════════╪══════════╡
│ 172.0HP 1.6L 4 Cylinder Engine… ┆ 172  ┆ 1.6          ┆ 4        ┆ false    │
│ 3.0L Twin Turbo                 ┆ null ┆ 3.0          ┆ null     ┆ false    │
│ 429.0HP 5.0L 8 Cylinder Engine… ┆ 429  ┆ 5.0          ┆ 8        ┆ false    │
└─────────────────────────────────┴──────┴──────────────┴──────────┴──────────┘

In [19]:
xd =(
    df
    .with_columns(
        pl.col("engine").str.replace(r'\sLiter', "L")
    )
    .with_columns(
        pl.col("engine").str.extract(r'(\d+)\.\d+HP',1).alias("HP"),
        pl.col("engine").str.extract(r'(\d+\.\d+)L',1).alias("Displacement"),
        pl.col("engine").str.extract(r'(\d+)\sCylinder',1).alias("Cylinder"),
        pl.col("engine").str.contains("Electric").alias("Electric")
    )
)
xd

shape: (188_533, 17)
┌────────┬──────────────┬──────────────┬────────────┬───┬──────┬─────────────┬──────────┬──────────┐
│ id     ┆ brand        ┆ model        ┆ model_year ┆ … ┆ HP   ┆ Displacemen ┆ Cylinder ┆ Electric │
│ ---    ┆ ---          ┆ ---          ┆ ---        ┆   ┆ ---  ┆ t           ┆ ---      ┆ ---      │
│ i64    ┆ str          ┆ str          ┆ i64        ┆   ┆ str  ┆ ---         ┆ str      ┆ bool     │
│        ┆              ┆              ┆            ┆   ┆      ┆ str         ┆          ┆          │
╞════════╪══════════════╪══════════════╪════════════╪═══╪══════╪═════════════╪══════════╪══════════╡
│ 0      ┆ MINI         ┆ Cooper S     ┆ 2007       ┆ … ┆ 172  ┆ 1.6         ┆ 4        ┆ false    │
│        ┆              ┆ Base         ┆            ┆   ┆      ┆             ┆          ┆          │
│ 1      ┆ Lincoln      ┆ LS V8        ┆ 2002       ┆ … ┆ 252  ┆ 3.9         ┆ 8        ┆ false    │
│ 2      ┆ Chevrolet    ┆ Silverado    ┆ 2002       ┆ … ┆ 320  ┆ 5.3         ┆ 8        ┆ false    │
│        ┆              ┆ 2500 LT      ┆            ┆   ┆      ┆             ┆          ┆          │
│ 3      ┆ Genesis      ┆ G90 5.0      ┆ 2017       ┆ … ┆ 420  ┆ 5.0         ┆ 8        ┆ false    │
│        ┆              ┆ Ultimate     ┆            ┆   ┆      ┆             ┆          ┆          │
│ 4      ┆ Mercedes-Ben ┆ Metris Base  ┆ 2021       ┆ … ┆ 208  ┆ 2.0         ┆ 4        ┆ false    │
│        ┆ z            ┆              ┆            ┆   ┆      ┆             ┆          ┆          │
│ …      ┆ …            ┆ …            ┆ …          ┆ … ┆ …    ┆ …           ┆ …        ┆ …        │
│ 188528 ┆ Cadillac     ┆ Escalade ESV ┆ 2017       ┆ … ┆ 420  ┆ 6.2         ┆ 8        ┆ false    │
│        ┆              ┆ Platinum     ┆            ┆   ┆      ┆             ┆          ┆          │
│ 188529 ┆ Mercedes-Ben ┆ AMG C 43 AMG ┆ 2018       ┆ … ┆ 385  ┆ 3.0         ┆ 6        ┆ false    │
│        ┆ z            ┆ C 43 4MATIC  ┆            ┆   ┆      ┆             ┆          ┆          │
│ 188530 ┆ Mercedes-Ben ┆ AMG GLC 63   ┆ 2021       ┆ … ┆ 469  ┆ 4.0         ┆ 8        ┆ false    │
│        ┆ z            ┆ Base 4MATIC  ┆            ┆   ┆      ┆             ┆          ┆          │
│ 188531 ┆ Audi         ┆ S5 3.0T      ┆ 2022       ┆ … ┆ null ┆ 3.0         ┆ null     ┆ false    │
│        ┆              ┆ Prestige     ┆            ┆   ┆      ┆             ┆          ┆          │
│ 188532 ┆ Porsche      ┆ Macan Base   ┆ 2016       ┆ … ┆ 252  ┆ 2.0         ┆ 4        ┆ false    │
└────────┴──────────────┴──────────────┴────────────┴───┴──────┴─────────────┴──────────┴──────────┘

In [27]:
(
    df
    .with_columns()
    .with_columns(set_brand(pl.col("brand")).alias("Brand Type"))
    .with_columns(set_fuel_type(pl.col("fuel_type")).alias("Fuel Type"))
)

shape: (188_533, 15)
┌────────┬─────────────┬────────────┬────────────┬───┬────────────┬───────┬────────────┬───────────┐
│ id     ┆ brand       ┆ model      ┆ model_year ┆ … ┆ clean_titl ┆ price ┆ Brand Type ┆ Fuel Type │
│ ---    ┆ ---         ┆ ---        ┆ ---        ┆   ┆ e          ┆ ---   ┆ ---        ┆ ---       │
│ i64    ┆ str         ┆ str        ┆ i64        ┆   ┆ ---        ┆ i64   ┆ i32        ┆ i32       │
│        ┆             ┆            ┆            ┆   ┆ str        ┆       ┆            ┆           │
╞════════╪═════════════╪════════════╪════════════╪═══╪════════════╪═══════╪════════════╪═══════════╡
│ 0      ┆ MINI        ┆ Cooper S   ┆ 2007       ┆ … ┆ Yes        ┆ 4200  ┆ 2          ┆ 1         │
│        ┆             ┆ Base       ┆            ┆   ┆            ┆       ┆            ┆           │
│ 1      ┆ Lincoln     ┆ LS V8      ┆ 2002       ┆ … ┆ Yes        ┆ 4999  ┆ 2          ┆ 1         │
│ 2      ┆ Chevrolet   ┆ Silverado  ┆ 2002       ┆ … ┆ Yes        ┆ 13900 ┆ 1          ┆ 1         │
│        ┆             ┆ 2500 LT    ┆            ┆   ┆            ┆       ┆            ┆           │
│ 3      ┆ Genesis     ┆ G90 5.0    ┆ 2017       ┆ … ┆ Yes        ┆ 45000 ┆ 2          ┆ 1         │
│        ┆             ┆ Ultimate   ┆            ┆   ┆            ┆       ┆            ┆           │
│ 4      ┆ Mercedes-Be ┆ Metris     ┆ 2021       ┆ … ┆ Yes        ┆ 97500 ┆ 2          ┆ 1         │
│        ┆ nz          ┆ Base       ┆            ┆   ┆            ┆       ┆            ┆           │
│ …      ┆ …           ┆ …          ┆ …          ┆ … ┆ …          ┆ …     ┆ …          ┆ …         │
│ 188528 ┆ Cadillac    ┆ Escalade   ┆ 2017       ┆ … ┆ Yes        ┆ 27500 ┆ 2          ┆ 1         │
│        ┆             ┆ ESV        ┆            ┆   ┆            ┆       ┆            ┆           │
│        ┆             ┆ Platinum   ┆            ┆   ┆            ┆       ┆            ┆           │
│ 188529 ┆ Mercedes-Be ┆ AMG C 43   ┆ 2018       ┆ … ┆ Yes        ┆ 30000 ┆ 2          ┆ 1         │
│        ┆ nz          ┆ AMG C 43   ┆            ┆   ┆            ┆       ┆            ┆           │
│        ┆             ┆ 4MATIC     ┆            ┆   ┆            ┆       ┆            ┆           │
│ 188530 ┆ Mercedes-Be ┆ AMG GLC 63 ┆ 2021       ┆ … ┆ Yes        ┆ 86900 ┆ 2          ┆ 1         │
│        ┆ nz          ┆ Base       ┆            ┆   ┆            ┆       ┆            ┆           │
│        ┆             ┆ 4MATIC     ┆            ┆   ┆            ┆       ┆            ┆           │
│ 188531 ┆ Audi        ┆ S5 3.0T    ┆ 2022       ┆ … ┆ null       ┆ 84900 ┆ 2          ┆ 1         │
│        ┆             ┆ Prestige   ┆            ┆   ┆            ┆       ┆            ┆           │
│ 188532 ┆ Porsche     ┆ Macan Base ┆ 2016       ┆ … ┆ Yes        ┆ 28995 ┆ 2          ┆ 1         │
└────────┴─────────────┴────────────┴────────────┴───┴────────────┴───────┴────────────┴───────────┘